lib/charms/layer/basic.py

```diff
@@ -324,6 +332,8 @@ def _strip_ext(s):
         for ending in [".zip", ".tar.gz"]:
             if s.endswith(ending):
                 return s[:-len(ending)]
+        if s.endswith('.whl'):
+            return s.split('-', 1)[0].replace('_', '-')
```

Comment:
I would probably create a small function that gets PEP 503 normalized names:

https://peps.python.org/pep-0503/#normalized-names
```
import re

def normalize(name):
    return re.sub(r"[-_.]+", "-", name).lower()
```

The binary distribution names should be normalized names with `-` replaced by `_` judging by: https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode

I think referring to PEP 503 would make it more clear why we are doing this replacement.

Reply:
> Looks good to me overall, although I would slightly refactor the `_add_back_versions` function modification.

Thanks for the review, Dima. Yes, I'll revisit that one.

To be honest I'm not quite sure why this is needed both in `_load_wheelhouse_versions` and  `_add_back_versions` as the second gets its input from the former, but I guess it was added for some reason that is not obvious so I'll keep it.

Moving this to a separate helper instead of duplicating the code would be good though.

The PEP 503 reference also makes sense, although I do not particularly like regular expressions when the same can be done with string manipulation.

Reply:
Ah, it appears to me that the original authors gave up on fully parsing the variety of source wheel formats out there by using the `LooseVersion` package, and just including whatever comes after the version string into the version comparison.

Example from the unit test:
```python3
    {
        'python-dateutil': LooseVersion ('2.8.1.tar.gz'),
        'setuptools-scm': LooseVersion ('1.17.0.tar.gz'),
        'wheel': LooseVersion ('0.33.6.tar.gz'),
    }
```

So this is probably why the second method exist.

Appears to me we should probably try to unpickle this a bit, but if we introduce a strict normalization we should probably contain that new handling to the binary wheels, so that we don't mess with existing in the wild setups using the source build method.

Reply:
> The binary distribution names should be normalized names with - replaced by _ judging by: https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode

Regarding this, the existing code is actually relying on the distributed wheels to already follow this normalization, and does the opposite and replaces `_` (LOW LINE) and stores the package names with `-` (HYPHEN) in its internal representation.

I see no need to change this, and the internal handling of both wheel types must be identical in order for upgrades to succeed.